Add ImageGrid widget

janhartman opened the pull request:
This commit adds a new widget, Image Grid - should eventually be moved to ImageAnalytics.
The widget still needs additional tests and documentation.

##### Description of changes
- adding a new widget

##### Includes
- [X] Code changes
- [ ] Tests
- [ ] Documentation

kernc commented on orangecontrib/prototypes/image_grid.py:
```diff
@@ -0,0 +1,148 @@
+        If the desired grid size is not provided, it is calculated from the data.
+        
+        :param size_x: The number of columns
+        :param size_y: The number of rows
```
In PyCharm, set [auto-docstring format to NumPy or Google style](https://www.jetbrains.com/help/pycharm/python-integrated-tools.html).

kernc commented on orangecontrib/prototypes/image_grid.py:
```diff
@@ -0,0 +1,148 @@
+        :param use_cosine: precompute cosine distances and pass them to MDS
+        :return: the data in 2 dimensions
+        """
+        if method is "MDS":
```
== to compare strings.

That `"MDS" is "MDS"` is just CPython implementation detail.

kernc commented on orangecontrib/prototypes/widgets/owimagegrid.py:
```diff
@@ -0,0 +1,1141 @@
+class OWImageGrid(widget.OWWidget):
+    name = "Image Grid"
+    description = "Visualize images in a similarity grid"
+    icon = "icons/Category-Prototypes.svg"
```
Propose an icon.

ajdapretnar replied:
Let me know the next time you are in the lab and I'll show you how to propose an icon.

janhartman replied:
Will do @ajdapretnar, thanks!

kernc commented on orangecontrib/prototypes/widgets/owimagegrid.py:
```diff
@@ -0,0 +1,1141 @@
+    description = "Visualize images in a similarity grid"
+    icon = "icons/Category-Prototypes.svg"
+    priority = 160
+    category = ""
```
strip.

kernc commented on orangecontrib/prototypes/widgets/owimagegrid.py:
```diff
@@ -0,0 +1,1141 @@
+
+    class Inputs:
+        data = Input("Embeddings", Orange.data.Table)
+        # data_subset = Input("Embeddings Subset", Orange.data.Table)
```
Don't leave in commented code without explaining what's it for.

kernc commented on orangecontrib/prototypes/widgets/owimagegrid.py:
```diff
@@ -0,0 +1,1141 @@
+    class Outputs:
+        data = Output("Images", Orange.data.Table)
+
+    _NO_DATA_INFO_TEXT = "No data on input."
```
Unused.

kernc commented on orangecontrib/prototypes/widgets/owimagegrid.py:
```diff
@@ -0,0 +1,1141 @@
+
+    _NO_DATA_INFO_TEXT = "No data on input."
+
+    settings_version = 2
```
1 or strip.

kernc commented on orangecontrib/prototypes/widgets/owimagegrid.py:
```diff
@@ -0,0 +1,1141 @@
+            self.controlArea, self, "imageAttr",
+            box="Image Filename Attribute",
+            tooltip="Attribute with image filenames",
+            callback=[self.clearScene, self.setupScene],
```
This actually works? Who would have thought.

janhartman replied:
Surprisingly, it does, this was already in Image Viewer.

kernc commented on orangecontrib/prototypes/widgets/owimagegrid.py:
```diff
@@ -0,0 +1,1141 @@
+            url = origin.resolved(name)
+        if not url.scheme():
+            url.setScheme("file")
+        return url
```
This looks messy. I posit the following snippet works the same for all inputs.
```py
def valueToQUrl(self, value):
    base = value.variable.attributes.get('origin', '')
    if QDir(base).exists():
        base = QUrl.fromLocalFile(base)
    else:
        base = QUrl(base)
    url = base.resolved(QUrl(value))
    return url
```
Don't we have this somewhere already?

janhartman replied:
Using this instead of the old function, thanks.

kernc commented on orangecontrib/prototypes/widgets/owimagegrid.py:
```diff
@@ -0,0 +1,1141 @@
+        self.clear()
+
+
+# Classes from Image Viewer, adapted with some new options.
```
What new options? Why must codes diverge?

kernc commented on orangecontrib/prototypes/widgets/owimagegrid.py:
```diff
@@ -0,0 +1,1141 @@
+def main(argv=None):
+    import sys
+    """
+    Will be uncommented when the widget is moved to ImageAnalytics.
```
Mark future work with 'TODO' so it is more easily spotted / IDE notifies about.

kernc replied:
Also why? If running the prototype file will work, it'll work; if it won't, it won't. With this commented out, it just never will.

janhartman replied:
I couldn't figure out a way to make the imports for ImportImages and ImageEmbedder work since they reside in the imageanalytics module. Pycharm did not help me either.

kernc replied:
I don't understand. Imports as you have them are correct. When imageanalytics add-on is properly installed (i.e. `pip install -e` if prototypes was installed with `-e`), they will resolve. In either event, what good is a main function that just returns?

kernc replied:
Consider constructing `main()` in such a way that someone who's a working installation of orange3-imageanalytics can run the module and see what's what-

janhartman replied:
The main function now runs properly. Looks like I had to reinstall ImageAnalytics.

janhartman commented:
I fixed most of the issues and added some comments in [commit bdeb6c4](https://github.com/biolab/orange3-prototypes/pull/101/commits/bdeb6c482bb3d72be3a447d5d3366eb71cf34ac1)

BlazZupan commented:
Icon is coming, most likely tomorrow (Friday).

kernc commented on requirements.txt:
```diff
@@ -0,0 +1 @@
+lap==0.3.0
```
This file is not used; add to _setup.py_.

kernc commented on orangecontrib/prototypes/widgets/owimagegrid.py:
```diff
@@ -0,0 +1,1135 @@
+        if pos is None:
+            return False
+
+        i, j = pos
```
This code above, is it the same as this?
```py
for i in range(layout.rowCount()):
    for j in range(layout.columnCount()):
        if layout.itemAt(i, j) == current:
            break
else:
    return False

...  # do something with i and j
```

kernc replied:
I see it's copied line for line from Image Viewer. Never mind. If it's good enough for main ...

kernc commented on orangecontrib/prototypes/widgets/owimagegrid.py:
```diff
@@ -0,0 +1,1135 @@
+                thumbnail = GraphicsThumbnailWidget(QPixmap(), crop=self.cell_fit == 1)
+                thumbnail.setThumbnailSize(size)
+                thumbnail.instance = inst
+                self.thumbnailView.addThumbnail(thumbnail)
```
How about if empty space weren't an item one can click/select?

janhartman replied:
fixed in [8f8c7c9](https://github.com/biolab/orange3-prototypes/pull/101/commits/8f8c7c930f785ef15c4da95d5210cbe40d3a69ad)

BlazZupan commented:
I am merging this one. Good enough for Prototypes. Everything works fine. Some minor improvements regarding selection are probably still needed but should be managed in a separate commit.
